File: Ejercicio 1/uber.py
from abc import ABC, abstractmethod

class Conduccion(ABC):
    @abstractmethod
    def iniciarViaje(self): 
        pass

class Transporte(ABC):
    def __init__(self, conductor):
        self.conductor = conductor

    @abstractmethod
    def tipo(self): 
        pass


class CarroUber(Transporte, Conduccion):
    def tipo(self): 
        print(f"Carro Uber de {self.conductor}")
        
    def iniciarViaje(self): 
        print("Viaje en carro iniciado")

class MotoUber(Transporte, Conduccion):
    def tipo(self): 
        print(f"Moto Uber de {self.conductor}")
        
    def iniciarViaje(self): 
        print("Viaje en moto iniciado")

class Bicicleta(Transporte, Conduccion):
    def tipo(self): 
        print(f"Bicicleta de {self.conductor}")
        
    def iniciarViaje(self): 
        print("Viaje en bici iniciado")


transportes = [
    CarroUber("Juan"),
    MotoUber("Ana"),
    Bicicleta("Luis")
]


for t in transportes:
    t.tipo()
    t.iniciarViaje()
    print("-" * 20)

File: Ejercicio 1/uber.java
interface Conduccion {
    void iniciarViaje();
}


abstract class Transporte implements Conduccion {
    String conductor;

    Transporte(String c) { this.conductor = c; }

    abstract void tipo();
    
}

class CarroUber extends Transporte {
    CarroUber(String c) { super(c); }

    void tipo() { System.out.println("Carro Uber de " + conductor); }
    @Override
    public void iniciarViaje() { System.out.println("Viaje en carro iniciado"); }
}

class MotoUber extends Transporte {
    MotoUber(String c) { super(c); }

    void tipo() { System.out.println("Moto Uber de " + conductor); }
    @Override
    public void iniciarViaje() { System.out.println("Viaje en moto iniciado"); }
}

class Bicicleta extends Transporte {
    Bicicleta(String c) { super(c); }

    void tipo() { System.out.println("Bicicleta de " + conductor); }
    @Override
    public void iniciarViaje() { System.out.println("Viaje en bici iniciado"); }
}

public class Main {
    public static void main(String[] args) {
        Transporte[] lista = {
            new CarroUber("Juan"),
            new MotoUber("Ana"),
            new Bicicleta("Luis")
        };

        for (Transporte t : lista) {
            t.tipo();
            t.iniciarViaje();
            System.out.println("---");
        }
    }
}
